File: Top 100 Codes/03-Codes for Recursion/18-Given a set of positive integers, find all its subsets –/Givenasetofpositiveintegersfindallitssubsets.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to print a subset
void printSubset(const vector<int>& subset) {
    cout << "{ ";
    for (int num : subset) {
        cout << num << " ";
    }
    cout << "}" << endl;
}

// Recursive function to generate all subsets
void generateSubsets(vector<int>& set, vector<int>& current, int index) {
    if (index == set.size()) {
        printSubset(current);  // Print the current subset
        return;
    }

    // Exclude the current element and move to the next
    generateSubsets(set, current, index + 1);

    // Include the current element in the subset
    current.push_back(set[index]);
    generateSubsets(set, current, index + 1);

    // Backtrack by removing the last element
    current.pop_back();
}

int main() {
    vector<int> set = {1, 2, 3};  // Example set
    vector<int> current;          // Current subset
    cout << "All subsets:" << endl;
    generateSubsets(set, current, 0);
    return 0;
}

File: Top 100 Codes/03-Codes for Recursion/18-Given a set of positive integers, find all its subsets –/Givenasetofpositiveintegersfindallitssubsets.py
def generate_subsets(nums, current, index):
    if index == len(nums):
        print(current)  # Print the current subset
        return

    # Exclude the current element
    generate_subsets(nums, current, index + 1)
    
    # Include the current element
    generate_subsets(nums, current + [nums[index]], index + 1)

# Example usage
nums = [1, 2, 3]  # Example set
print("All subsets:")
generate_subsets(nums, [], 0)
